src/backend/distributed/README.md

```diff
@@ -1955,12 +1955,153 @@ The distributed transactionId and backend/PID mapping is done via BackendData st
+  ....
+
+  DETAIL:  on server onderkalaci@localhost:9702 connectionId: 2
+NOTICE:  issuing SELECT lock_shard_metadata(7, ARRAY[102008])
+DETAIL:  on server onderkalaci@localhost:9701 connectionId: 1
+NOTICE:  issuing SELECT lock_shard_metadata(7, ARRAY[102008])
+DETAIL:  on server onderkalaci@localhost:9702 connectionId: 2
+...
+```
+
+Another useful application of advisory locks in query from any node is modifications to reference tables. Reference tables (or in general replicated tables) have multiple placements for a given shard. So, modifications to the placements of the same shard should be serialized. We cannot allow multiple modification commands to modify the placements in different orders. It could cause diverging the contents of the data.
+
+The coordinator already serializes the modifications to reference tables via `LockShardResource()` C function. When there are other nodes in the cluster, Citus sends the similar command, to the first worker node in the cluster. In general, Citus aims to serialize operations on the reference tables via acquiring advisory locks on the first worker node, see `SerializeNonCommutativeWrites()` and `LockShardListResourcesOnFirstWorker()` C functions. The outcome is the following:
+
+```sql
+ BEGIN;
+BEGIN
+-- 9701 is the first worker node in the metadata
+insert into reference_table VALUES (1);
+NOTICE:  issuing BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED;SELECT assign_distributed_transaction_id(0, 8, '2023-09-18 16:30:18.715259+03');
+DETAIL:  on server onderkalaci@localhost:9701 connectionId: 1
+NOTICE:  issuing SELECT lock_shard_resources(3, ARRAY[102040])
+DETAIL:  on server onderkalaci@localhost:9701 connectionId: 1
+NOTICE:  issuing INSERT INTO public.reference_table_102040 (a) VALUES (1)
+
+``` 
+
+
+4. **Low-level Locks (SpinLocks and LWLocks)**: 
+
+SpinLocks and LWLocks are low-level locking mechanisms often used internally by the database system. 
+Citus uses `LwLocks` and `SpinLocks` as described in Postgres' source code: https://github.com/postgres/postgres/tree/master/src/backend/storage/lmgr
+
++ **Spinlocks:** Spinlocks are used for very short-term locking and are meant to be held only for a few instructions. They don't have features like deadlock detection or automatic release, and waiting processes keep checking ("busy-loop") until they can acquire the lock. In the past we had some bugs where we had a `palloc` failures while holding `SpinLock` and which prevented the lock to be released. So, be extra careful when using `SpinLock`.
+
++ **Lightweight Locks (LWLocks):** These locks are mainly used for controlling access to shared memory data structures and support both exclusive and shared lock modes. While they also lack deadlock detection, they do automatically release when errors are raised, and waiting processes block on a SysV semaphore, conserving CPU time.
+
+
```

Comment:
I think another thing that might be nice to discuss is locking around co-location (if you can figure out how it works now :D)

Reply:
added the logic behind `AcquireRebalanceColocationLock()` and `AcquirePlacementColocationLock()`